contracts/helpers/GenericHarvester.sol

```diff
@@ -0,0 +1,312 @@
+
+        _adjustAllowance(tokenIn, address(transmuter), amountOut);
+        uint256 amountStableOut = transmuter.swapExactInput(
+            amountOut,
+            minAmountOut,
+            tokenIn,
+            address(agToken),
+            address(this),
+            block.timestamp
+        );
+        if (amount > amountStableOut) {
+            if (budget[sender] < amount - amountStableOut) revert InsufficientFunds();
+            budget[sender] -= amount - amountStableOut;
+        }
+        return CALLBACK_SUCCESS;
+    }
+
+    /**
+     * @notice Add budget to a receiver
+     * @param amount amount of AGToken to add to the budget
+     * @param receiver address of the receiver
+     */
+    function addBudget(uint256 amount, address receiver) public virtual {
+        IERC20(agToken).safeTransferFrom(msg.sender, address(this), amount);
+
+        budget[receiver] += amount;
+    }
+
+    /**
+     * @notice Remove budget from a receiver
+     * @param amount amount of AGToken to remove from the budget
+     * @param receiver address of the receiver
+     */
+    function removeBudget(uint256 amount, address receiver) public virtual {
+        if (budget[msg.sender] < amount) revert InsufficientFunds();
+        budget[msg.sender] -= amount;
+
+        IERC20(agToken).safeTransfer(receiver, amount);
+    }
+
+    function _swapToTokenOut(
+        uint256 typeAction,
+        address tokenIn,
+        address tokenOut,
+        uint256 amount,
+        SwapType swapType,
+        bytes memory callData
+    ) internal returns (uint256 amountOut) {
+        if (swapType == SwapType.SWAP) {
+            amountOut = _swapToTokenOutSwap(tokenIn, tokenOut, amount, callData);
+        } else if (swapType == SwapType.VAULT) {
+            amountOut = _swapToTokenOutVault(typeAction, tokenOut, amount);
+        }
+    }
+
+    /**
+     * @notice Swap token using the router/aggregator
+     * @param tokenIn address of the token to swap
+     * @param tokenOut address of the token to receive
+     * @param amount amount of token to swap
+     * @param callData bytes to call the router/aggregator
+     */
+    function _swapToTokenOutSwap(
+        address tokenIn,
+        address tokenOut,
+        uint256 amount,
+        bytes memory callData
+    ) internal returns (uint256) {
+        uint256 balance = IERC20(tokenOut).balanceOf(address(this));
+
+        address[] memory tokens = new address[](1);
+        tokens[0] = tokenIn;
+        bytes[] memory callDatas = new bytes[](1);
+        callDatas[0] = callData;
+        uint256[] memory amounts = new uint256[](1);
+        amounts[0] = amount;
+        _swap(tokens, callDatas, amounts);
+
+        uint256 amountOut = IERC20(tokenOut).balanceOf(address(this)) - balance;
+        uint256 decimalsTokenIn = IERC20Metadata(tokenIn).decimals();
+        uint256 decimalsTokenOut = IERC20Metadata(tokenOut).decimals();
+
+        if (decimalsTokenIn > decimalsTokenOut) {
```

Comment:
let's internalize the function and put it in BaseHarvester we have a similar logic as well in MultiBlockHarvester when we compare decimalsAsset > 18.
Better to have it only in one place as an internal function